renderdoc/driver/d3d8/d3d8_manager.h

```diff
@@ -0,0 +1,101 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#pragma once
+
+#include "api/replay/renderdoc_replay.h"
+#include "common/wrapped_pool.h"
+#include "core/core.h"
+#include "core/resource_manager.h"
+#include "serialise/serialiser.h"
+#include "d3d8_common.h"
+
+struct D3D8ResourceRecord : public ResourceRecord
+{
+  enum
+  {
+    NullResource = NULL
+  };
+
+  D3D8ResourceRecord(ResourceId id) : ResourceRecord(id, true) {}
+  void Insert(map<int32_t, Chunk *> &recordlist)
+  {
+    bool dataWritten = DataWritten;
+
+    DataWritten = true;
+
+    for(auto it = Parents.begin(); it != Parents.end(); ++it)
+    {
+      if(!(*it)->DataWritten)
+      {
+        (*it)->Insert(recordlist);
+      }
+    }
+
+    if(!dataWritten)
+      recordlist.insert(m_Chunks.begin(), m_Chunks.end());
+  }
+};
+
+class D3D8ResourceManager
+    : public ResourceManager<IDirect3DResource8 *, IDirect3DResource8 *, D3D8ResourceRecord>
+{
+public:
+  D3D8ResourceManager(LogState state, Serialiser *ser, WrappedD3DDevice8 *dev)
+      : ResourceManager(state, ser), m_Device(dev)
+  {
+  }
+
+  IDirect3DResource8 *UnwrapResource(IDirect3DResource8 *res);
+
+private:
+  bool SerialisableResource(ResourceId id, D3D8ResourceRecord *record);
+  ResourceId GetID(IDirect3DResource8 *res);
+
+  bool ResourceTypeRelease(IDirect3DResource8 *res);
+
+  bool Force_InitialState(IDirect3DResource8 *res, bool prepare);
+  bool Need_InitialStateChunk(IDirect3DResource8 *res);
+  bool Prepare_InitialState(IDirect3DResource8 *res);
+  bool Serialise_InitialState(ResourceId resid, IDirect3DResource8 *res);
+  void Create_InitialState(ResourceId id, IDirect3DResource8 *live, bool hasData);
+  void Apply_InitialState(IDirect3DResource8 *live, InitialContentData data);
+
+  WrappedD3DDevice8 *m_Device;
+};
+
+template <typename Dest>
+typename Dest::InnerType *Unwrap(typename Dest::InnerType *obj)
+{
+#if ENABLED(RDOC_DEVEL)
+  if(obj && !Dest::IsAlloc(obj))
+  {
+    RDCERR("Trying to unwrap invalid type");
+    return NULL;
+  }
+#endif
+  return obj == NULL ? NULL : (Dest::InnerType *)((Dest *)obj)->GetReal();
+}
+
+#define UNWRAP(type, obj) Unwrap<type>((type::InnerType *)obj)
```

Comment:
Ouch, that is really ugly. Technically `DWORD` isn't even pointer sized, although I guess there's not even a 64-bit D3D8 runtime so it's moot :smile:.

I think the solution here would be to create your own interface called `IDirect3DShader8` and pretend that the `DWORD` is a pointer to it, then wrap that pretend interface. You might need to special case in a couple of places so that it doesn't try to use the 'real' pointer as `IUnknown` and call `AddRef/Release` on it, but otherwise you can just have it store the `DWORD` and pass that along to the API when necessary.

I think that should work? There might be snags when you go to implement it but that's probably the least-invasive way to massage it into working.